### PR TITLE
Patch to allow manual specification of dependencies

### PR DESCRIPTION
**Pull request by recheliu**
This PR will extend this branch to allow the manual specification of dependencies. It also fixes issues to link the code and run unit tests.

Detailed description of the changed files.

Modified CMakeLists.txt

```
[ADD] Define CMAKE_BUILD (or the exported symbols won't be included).
[ADD] When using MSVC, add a cache variable WITH_PREBUILD_MSVC_DEPENDENCIES.
```

Modified cmake/Dependencies.cmake

```
[MOD] For boost, do not append the auto-detected boost libraries when using manually-built 3rd-party dependencies.
[ADD] Add definition BOOST_NO_CXX11_TEMPLATE_ALIASES, or Visual Studio might crash.
[ADD] Allow the separation specification of szip and zlib.
[ADD] Require the C component from HDF5.
```

Modified cmake/Templates/caffe_config.h.in

```
[DEL] Remove already-defined preprocessors.
[MOD] Correct the example path.
```

Modified cmake/WindowsCreateLinkHeader.cmake

```
[MOD] When searching for .obj, using the native path.
```

Modified cmake/WindowsUtils.cmake

```
[MOD] Use the new variable WITH_PREBUILD_MSVC_DEPENDENCIES to skip the auto-downloaded dependencies.
```

Modified src/caffe/test/CMakeLists.txt

```
[MOD] Also build the solver data list.
```

Modified cmake/Modules/FindGlog.cmake

```
[MOD] Fix the search of glog headers/libraries under windows.
[ADD] Add a variable GLOG_STATIC for static-linking glog library.
```

Tests
1. Run on ubuntu. All unit test PASSED.
2. Run on Windows with visual studio 2013 as the generator. The dependencies were built by me. Use static linking. All unit test PASSED.
3. Run on Windows with visual studio 2013 as the generator. The dependencies were automatically downloaded. Use static linking. All unit test PASSED.
4. Run on Windows with the original .sln files.

### CODE REVIEW

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -4,7 +4,19 @@ set(Caffe_LINKER_LIBS "")
 # ---[ Boost
 find_package(Boost 1.46 REQUIRED COMPONENTS system thread filesystem)
 include_directories(SYSTEM ${Boost_INCLUDE_DIR})
-list(APPEND Caffe_LINKER_LIBS ${Boost_LIBRARIES})
+link_directories(
+    ${Boost_LIBRARY_DIR_RELEASE}
+)
+
+if(NOT DEFINED MSVC OR WITH_PREBUILD_MSVC_DEPENDENCIES )
```
Comment (willyd) [inline cmake/Dependencies.cmake]: I think we should avoid adding the link directories as this is considered bad practice in CMake. We should append Boost libraries to the list and add the BOOST_ALL_NO_LIB define.

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -23,8 +35,47 @@ list(APPEND Caffe_LINKER_LIBS ${GFLAGS_LIBRARIES})
 # ---[ Google-protobuf
 include(cmake/ProtoBuf.cmake)
 
+if( MSVC AND NOT WITH_PREBUILD_MSVC_DEPENDENCIES )
```
Comment (willyd) [inline cmake/Dependencies.cmake]: ZLIB and SZIP are not dependencies of Caffe per se but transitive dependencies of HDF5. This should be handled in a HDF5Config.cmake file or a custom FindHDF5.cmake module.

**Comment by willyd** [comment]
> [ADD] Define CMAKE_BUILD (or the exported symbols won't be included).

How is this necessary? Am I missing something here?

File: CMakeLists.txt [hunk 3]
```diff
@@ -14,7 +14,17 @@ set(CAFFE_TARGET_VERSION "1.0.0-rc3")
 set(CAFFE_TARGET_SOVERSION "1.0.0-rc3")
 add_definitions(-DCAFFE_VERSION=${CAFFE_TARGET_VERSION})
 
+# definition needed to include CMake generated files
+add_definitions(-DCMAKE_BUILD)
```
Comment (recheliu) [inline CMakeLists.txt:18]: This is needed for static linking. Otherwise, common.hpp will skip export.hpp, which consequently miss the required .obj. 

This is not an issue in linking time, but it will be an issue in runtime. You will get an error like this:

F0509 14:41:28.143733 47048 layer_factory.cpp:62] Check failed: registry.count(type) == 1 (0 vs. 1)
Unknown layer type: Input (known types: AbsVal, Accuracy, ArgMax, BNLL, BatchNorm, BatchReindex, Bia
s, Concat, ContrastiveLoss, Convolution, Crop, Data, Deconvolution, Dropout, DummyData, ELU, Eltwise
, Embed, EuclideanLoss, Exp, Filter, Flatten, HDF5Data, HDF5Output, HingeLoss, Im2col, ImageData, In
fogainLoss, InnerProduct, LRN, Log, MVN, MemoryData, MultinomialLogisticLoss, PReLU, Pooling, Power,
 ReLU, Reduction, Reshape, SPP, Scale, Sigmoid, SigmoidCrossEntropyLoss, Slice, Softmax, SoftmaxWith
Loss, Split, TanH, Threshold, Tile)

**Comment by willyd** [comment]
@recheliu You are right. Thanks for clarifying the code **I** wrote. :)

**Comment by recheliu** [comment]
Hi,

Just curious whether my PR requires more revision. 

One issue is that failed build with appveyor. However, as I did not change any settings, so far I haven't figured out the root cause.

Besides, regarding the merge to caffe's windows branch, do you that which criteria we should meet?

**Comment by willyd** [comment]
> One issue is that failed build with appveyor. However, as I did not change any settings, so far I haven't figured out the root cause.

I have experienced the same issue did not find any reason why the build fails. I beleive that this [commit](https://github.com/BVLC/caffe/commit/a473c7743331feaec3ee8b88646f81afff638bd0) might fix it. I will soon update my fork and see.

I see that you revised your PR based on my comments. Most of it looks good. The only thing I think should not be included is the FindHDF5Filters.cmake. I would recommend you take the same approach I used: Write your own FindHDF5.cmake file that calls into FindHDF5Filters.cmake and add the path containing these files to the CMAKE_MODULE_PATH via a cache file with `-C` option to cmake. See https://cmake.org/cmake/help/v3.5/manual/cmake.1.html.

Also, shouldn't the HDF5 cmake build provide the imported targets for ZLIB and SZIP in the hdf5-config.cmake? Have you checked this? If that is the case your FindHDF5.cmake would only need to call find_package with the NO_MODULE option.

What do you think?

**Comment by recheliu** [comment]
@willyd Thanks for your feedback. The background is that when I built the static libraries of both OpenCV and HDF5, both require zlib. As I don't want OpenCV to depend on HDF5's own zlib, I built zlib separately. In order to do so, I turned off HDF5_PACKAGE_EXTLIBS in HDF5's cmake settings, and thus I also built szip separately. That's why I have the extra settings for zlib and szip. 

I like the idea to use NO_MODULE. I will test it.

**Comment by recheliu** [comment]
I move the all logics to configure HDF5 to cmake/External/hdf5.cmake. I did not create FindHDF5.cmake though because I suppose that this file name should be reserved for HDF5 installation, not be used by the application side. Thus I take the patterns to configure gflags/glog instead.

**Comment by recheliu** [comment]
Just curious whether you have any comments for the changes. Actually the master branch of Caffe has included several bug fixes. I wish that we can merge these changes in to this branch soon so my colleagues can use it on Windows. Thanks!

**Comment by willyd** [comment]
Hi @recheliu, sorry I did not have time to merge these changes. Although I am not totally in agreement about the external/hdf5.cmake which in my opinion was intended to download and build the dependencies using ExternalProject_Add, I am ok with the rest of your changes so I will merge them soon. Since the official devs of the unofficial windows branch have not shown any interest in this cmake port I will probably diverge from the windows branch and merge back the master branch changes soon too. I am mostly interested in the RNN stuff that was commited recently. I will let you know whenever the changes are integrated. Sorry for the delay.

**Comment by recheliu** [comment]
I just realized that this 28 days already passed. I did the merge on my local machine, which did not see any issues. Thus I think that it should be very straightforward.

**Comment by willyd** [comment]
You are right I meant sooner than this when I  said I would merge them soon. I have caught up on some other urgent matters and now going off from work for 3 weeks. Sorry for the delay. I will merge your changes first thing when I get back.

**Comment by recheliu** [comment]
Hi,

Just realized that almost another 4 weeks passed. May I know more about your plan? 

Also, do you have any plan to revise the windows branch of the original Caffe? I really wish that all build system can be merged into cmake, which save lots of troubles. Thanks.

Best,

**Comment by willyd** [comment]
@recheliu Just got back from vacations yesterday. I will try to merge your PR this week.

> Also, do you have any plan to revise the windows branch of the original Caffe?

I was thinking on merging my branch with master since the windows branch has not been updated for some time. I might just make a second branch so that the windows branch does not get to cluttered. Also, I did not receive any feedback from Microsoft about my PR there does not seem to be a lot of interest in a CMake build system from their part. Or maybe they are just too busy to look at this too.

> I really wish that all build system can be merged into cmake, which save lots of troubles. Thanks.

I totally agree.

I'll keep you posted whenever I merge your PR.

**Comment by willyd** [comment]
@recheliu I just merged your changes. Please note that I removed the automatically downloaded dependencies part as I have switched to using conda to manage the dependencies. Sorry for the delay.
